core/domain/exp_services.py

```diff
@@ -455,160 +455,268 @@ def apply_change_list(
 
     Raises:
         Exception. Any entries in the changelist are invalid.
+        Exception. Solution cannot exist with None interaction id.
     """
     exploration = exp_fetchers.get_exploration_by_id(exploration_id)
     try:
         to_param_domain = param_domain.ParamChange.from_dict
         for change in change_list:
             if change.cmd == exp_domain.CMD_ADD_STATE:
-                exploration.add_states([change.state_name])
+                # Here we use cast because we are narrowing down the type from
+                # ExplorationChange to a specific change command.
+                add_state_cmd = cast(
+                    exp_domain.AddExplorationStateCmd,
+                    change
+                )
+                exploration.add_states([add_state_cmd.state_name])
             elif change.cmd == exp_domain.CMD_RENAME_STATE:
+                # Here we use cast because we are narrowing down the type from
+                # ExplorationChange to a specific change command.
+                rename_state_cmd = cast(
+                    exp_domain.RenameExplorationStateCmd,
+                    change
+                )
                 exploration.rename_state(
-                    change.old_state_name, change.new_state_name)
+                    rename_state_cmd.old_state_name,
+                    rename_state_cmd.new_state_name
+                )
             elif change.cmd == exp_domain.CMD_DELETE_STATE:
-                exploration.delete_state(change.state_name)
+                # Here we use cast because we are narrowing down the type from
+                # ExplorationChange to a specific change command.
+                delete_state_cmd = cast(
+                    exp_domain.DeleteExplorationStateCmd,
+                    change
+                )
+                exploration.delete_state(delete_state_cmd.state_name)
             elif change.cmd == exp_domain.CMD_EDIT_STATE_PROPERTY:
                 state: state_domain.State = exploration.states[
                     change.state_name]
-                if (change.property_name ==
+                # Here we use cast because we are narrowing down the type from
+                # ExplorationChange to a specific change command.
+                edit_state_property_cmd = cast(
+                    exp_domain.EditExplorationStatePropertyCmd,
+                    change
+                )
+                if (edit_state_property_cmd.property_name ==
                         exp_domain.STATE_PROPERTY_PARAM_CHANGES):
-                    # Here change is an instance of ExplorationChange and every
-                    # attribute on ExplorationChange is defined dynamically, and
-                    # every dynamically defined attribute have str type. Thus to
-                    # convert the type to list, we have used assert here.
-                    assert isinstance(change.new_value, list)
+                    # Here we use cast because in this 'if clause' we are
```

Comment:
Same notes as before about casting edit_state_property_cmd instead, and using subtypes.

I think the general rule is that you should only cast to a explicit subtype.